src/service.js

```diff
@@ -238,8 +268,29 @@ export function getServiceForDoc(elementOrAmpDoc, id) {
  * If service `id` is not registered, returns null.
  * @param {!Element|!ShadowRoot} element
  * @param {string} id
+ * @param {boolean=} opt_searchElementPath If true, searches for element-held
+ *   services between `element` and its root node. This is a special case that
+ *   allows services to be scoped to an element subtree rather than the
+ *   enclosing AmpDoc.
+ * @return {?T}
+ * @template T
  */
-function getServiceForDocOrNullInternal(element, id) {
+function getServiceForDocOrNullInternal(element, id, opt_searchElementPath) {
+  // This is a first step in decoupling services from AmpDoc. It allows
+  // AMP elements to override services in the scope of their DOM subtrees,
+  // e.g. amp-script has special behavior for URL replacements for its children.
+  if (opt_searchElementPath) {
+    for (
+      let e = element;
+      e;
+      e = closestAncestorElementBySelector(e, '.i-amphtml-element')
```

Comment:
It might be faster if we could do a special class or attribute? E.g. if `url-replacements` is set on some element, we just do `x.closest('[i-amphtml-service-url-replacements]')`?

Reply:
Interesting idea, but I'd prefer to defer changing DOM until we're sure the optimization makes a difference. E.g. I don't think we have AMP element nesting depth > 3 in practice.

Reply:
This is currently a rare case so it's ok. But in our tests anything more than 1 parent up was showing benefits on `closest()`.